## Pandas

### Using the Pandas Python Library
Download data using requests 

In [1]:
from IPython.display import Image

In [2]:
import requests

download_url = "https://raw.githubusercontent.com/fivethirtyeight/data/master/nba-elo/nbaallelo.csv"
target_csv_path = "nba_all_elo.csv"

response = requests.get(download_url)
response.raise_for_status()    # Check that the request was successful
with open(target_csv_path, "wb") as f:
    f.write(response.content)
print("Download ready.")

Download ready.


* When you execute the script, it will save the file nba_all_elo.csv in your current working directory.
<br>
Download script has several advantages:
* You can tell where you got your data.
* You can repeat the download anytime! That’s especially handy if the data is often refreshed.
* You don’t need to share the 17MB CSV file with your co-workers. Usually, it’s enough to share the download script.
<br>
Now you can use the Pandas Python library to take a look at your data:



In [23]:
import pandas as pd 
df= pd.read_csv("nba_all_elo.csv")

In [24]:
# check the type 
type(df)

pandas.core.frame.DataFrame

Pandas Python library provides several similar functions like read_json(), read_html(), and read_sql_table()
<br>
How much data nba contains

In [5]:
len(df)

126314

In [6]:
df.shape[1]

23

* Python built-in function len() to determine the number of rows. 
* You also use the .shape attribute of the DataFrame to see its dimensionality. The result is a tuple containing the number of rows and columns

In [28]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts    elo_i    elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.00  1293.28      40.29    NYK   
1            0     NYK   Knicks   68  1300.00  1306.72      41.71    TRH   
2            0     CHS    Stags   63  1300.00  1309.65      42.01    NYK   
3            0     NYK   Knicks   47  1306.72  1297.07      40.69    CHS   
4            0     DTF  Falcons   33  1300.00  1279.62      38.86    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68    1300.00    1306.72             H           L   
1   Huskies       66    1300.00    1293.28             A           W   
2    Knicks       47    1306.72    1297.07             H           W   
3     Stags       63    1300.00    1309.65             A           L   
4  Capitols       50    1300.00    1320.38             H           L   

   forecast notes  
0      0.64   NaN  
1      0.36   NaN  
2      0.63   NaN  
3      0.37   NaN  
4      0.64   NaN

In [8]:
df.tail()

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
126309         100            1     CLE  Cavaliers  ...  60.309792     GSW   
126310         102            1     GSW   Warriors  ...  68.013329     CLE   
126311         101            1     CLE  Cavaliers  ...  60.010067     GSW   
126312         102            1     CLE  Cavaliers  ...  59.290245     GSW   
126313         103            1     GSW   Warriors  ...  68.519516     CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  \
126309   Warriors      103  1790.9591  1809.9791              H            L   
126310  Cavaliers       91  1704.3949  1700.7391              H            W   
126311   Warriors      104  1809.9791  1813.6349              A            L   
126312   Warriors      105  1813.6349  1822.2881              H            L   
126313  Cavaliers       97  1700.7391  1692.0859              A            W   

        forecast notes  
126309  0.546572   NaN  
126310  0.765565   NaN  
126311  0.234435   NaN  
126312  0.481450   NaN  
126313  0.518550   NaN  

[5 rows x 23 columns]

* See a column of ellipses (...) indicating the missing data. 
* If you’re working in a terminal, then that’s probably more readable than wrapping long rows. 
* You can configure Pandas to display all 23 columns like this

In [9]:
# Display all 
pd.set_option("display.max.columns", None)

In [10]:
# won’t need six decimal places! Change it to two
pd.set_option("display.precision", 2)


In [27]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts    elo_i    elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.00  1293.28      40.29    NYK   
1            0     NYK   Knicks   68  1300.00  1306.72      41.71    TRH   
2            0     CHS    Stags   63  1300.00  1309.65      42.01    NYK   
3            0     NYK   Knicks   47  1306.72  1297.07      40.69    CHS   
4            0     DTF  Falcons   33  1300.00  1279.62      38.86    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68    1300.00    1306.72             H           L   
1   Huskies       66    1300.00    1293.28             A           W   
2    Knicks       47    1306.72    1297.07             H           W   
3     Stags       63    1300.00    1309.65             A           L   
4  Capitols       50    1300.00    1320.38             H           L   

   forecast notes  
0      0.64   NaN  
1      0.36   NaN  
2      0.63   NaN  
3      0.37   NaN  
4      0.64   NaN

## Getting to Know Your Data
How to examine your data?
#### Displaying Data Types


In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   gameorder      126314 non-null  int64  
 1   game_id        126314 non-null  object 
 2   lg_id          126314 non-null  object 
 3   _iscopy        126314 non-null  int64  
 4   year_id        126314 non-null  int64  
 5   date_game      126314 non-null  object 
 6   seasongame     126314 non-null  int64  
 7   is_playoffs    126314 non-null  int64  
 8   team_id        126314 non-null  object 
 9   fran_id        126314 non-null  object 
 10  pts            126314 non-null  int64  
 11  elo_i          126314 non-null  float64
 12  elo_n          126314 non-null  float64
 13  win_equiv      126314 non-null  float64
 14  opp_id         126314 non-null  object 
 15  opp_fran       126314 non-null  object 
 16  opp_pts        126314 non-null  int64  
 17  opp_elo_i      126314 non-nul

* Pandas uses the NumPy library to work with these types
* The object data type is a special one. According to the Pandas Cookbook, the object data type is “a catch-all for columns that Pandas doesn’t recognize as any other specific type.” In practice, it often means that all of the values in the column are strings.

### Showing Basics Statistics

In [31]:
df.describe()

gameorder   _iscopy    year_id  seasongame  is_playoffs        pts  \
count  126314.00  126314.0  126314.00   126314.00    126314.00  126314.00   
mean    31579.00       0.5    1988.20       43.53         0.06     102.73   
std     18231.93       0.5      17.58       25.38         0.24      14.81   
min         1.00       0.0    1947.00        1.00         0.00       0.00   
25%     15790.00       0.0    1975.00       22.00         0.00      93.00   
50%     31579.00       0.5    1990.00       43.00         0.00     103.00   
75%     47368.00       1.0    2003.00       65.00         0.00     112.00   
max     63157.00       1.0    2015.00      108.00         1.00     186.00   

           elo_i      elo_n  win_equiv    opp_pts  opp_elo_i  opp_elo_n  \
count  126314.00  126314.00  126314.00  126314.00  126314.00  126314.00   
mean     1495.24    1495.24      41.71     102.73    1495.24    1495.24   
std       112.14     112.46      10.63      14.81     112.14     112.46   
min      1091.64    1085.77      10.15       0.00    1091.64    1085.77   
25%      1417.24    1416.99      34.10      93.00    1417.24    1416.99   
50%      1500.95    1500.95      42.11     103.00    1500.95    1500.95   
75%      1576.06    1576.29      49.64     112.00    1576.06    1576.29   
max      1853.10    1853.10      71.11     186.00    1853.10    1853.10   

        forecast  
count  126314.00  
mean        0.50  
std         0.22  
min         0.02  
25%         0.33  
50%         0.50  
75%         0.67  
max         0.98

describe() only analyzes numeric columns by default, but you can provide other data types if you use the include parameter

In [33]:
df.describe(include=object)


game_id   lg_id  date_game team_id fran_id  opp_id opp_fran  \
count         126314  126314     126314  126314  126314  126314   126314   
unique         63157       2      12426     104      53     104       53   
top     194611010TRH     NBA  4/13/2011     BOS  Lakers     BOS   Lakers   
freq               2  118016         30    5997    6024    5997     6024   

       game_location game_result           notes  
count         126314      126314            5424  
unique             3           2             231  
top                H           L  at New York NY  
freq           63138       63157             440

.describe() won’t try to calculate a mean or a standard deviation for the object columns, since they mostly include text strings. However, it will still display some descriptive statistics:



### Exploring Your Dataset


In [34]:
# list[""]
df["team_id"].value_counts()

BOS    5997
NYK    5769
LAL    5078
DET    4985
PHI    4533
       ... 
INJ      60
PIT      60
DTF      60
TRH      60
SDS      11
Name: team_id, Length: 104, dtype: int64

In [35]:
df["fran_id"].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Waterloo     

It seems that a team named "Lakers" played 6024 games, but only 5078 of those were played by the Los Angeles Lakers. Find out who the other "Lakers" team is:



In [125]:
df.loc[df["fran_id"] == "Lakers","pts"]
# df['pts'].valeus_counts()

1136       84
1152       62
1159       55
1170       77
1183       90
         ... 
126016    101
126052    106
126086    106
126115     92
126137     99
Name: pts, Length: 6024, dtype: int64

In [36]:
df.loc[df["fran_id"] == "Lakers", "team_id"].value_counts()

LAL    5078
MNL     946
Name: team_id, dtype: int64

* Lakers ("MNL") played 946 games. You can even find out when they played those games. 
* For that, you’ll first define a column that converts the value of date_game to the datetime data type. 
* Then you can use the min and max aggregate functions, to find the first and last games of Minneapolis Lakers

In [30]:
df["date_played"] = pd.to_datetime(df["date_game"])
df.loc[df["team_id"] == "MNL", "date_played"].min()

Timestamp('1948-11-04 00:00:00')

In [38]:
df.loc[df['team_id'] == 'MNL', 'date_played'].max()

Timestamp('1960-03-26 00:00:00')

In [31]:
df_test

gameorder       game_id lg_id  _iscopy  year_id   date_game  seasongame  \
1136        569  194811040BLB   NBA        1     1949   11/4/1948           1   
1152        577  194811060WSC   NBA        1     1949   11/6/1948           2   
1159        580  194811090BOS   NBA        1     1949   11/9/1948           3   
1170        586  194811110NYK   NBA        1     1949  11/11/1948           4   
1183        592  194811130PRO   NBA        1     1949  11/13/1948           5   
...         ...           ...   ...      ...      ...         ...         ...   
9701       4851  196003190LAL   NBA        0     1960   3/19/1960          80   
9702       4852  196003200LAL   NBA        0     1960   3/20/1960          81   
9709       4855  196003220STL   NBA        1     1960   3/22/1960          82   
9711       4856  196003240LAL   NBA        0     1960   3/24/1960          83   
9715       4858  196003260STL   NBA        1     1960   3/26/1960          84   

      is_playoffs team_id fran_id  pts    elo_i    elo_n  win_equiv opp_id  \
1136            0     MNL  Lakers   84  1527.00  1542.25      51.22    BLB   
1152            0     MNL  Lakers   62  1542.25  1536.21      50.62    WSC   
1159            0     MNL  Lakers   55  1536.21  1513.49      48.34    BOS   
1170            0     MNL  Lakers   77  1513.49  1524.96      49.50    NYK   
1183            0     MNL  Lakers   90  1524.96  1533.27      50.33    PRO   
...           ...     ...     ...  ...      ...      ...        ...    ...   
9701            1     MNL  Lakers   89  1471.07  1464.61      36.05    STL   
9702            1     MNL  Lakers  103  1464.61  1469.56      36.57    STL   
9709            1     MNL  Lakers  117  1469.56  1484.67      38.14    STL   
9711            1     MNL  Lakers   96  1484.67  1465.59      36.15    STL   
9715            1     MNL  Lakers   86  1465.59  1461.14      35.69    STL   

          opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
1136     Baltimore       72    1507.21    1491.95             A           W   
1152      Capitols       67    1481.87    1487.92             A           L   
1159       Celtics       77    1359.33    1382.05             A           L   
1170        Knicks       68    1460.42    1448.95             A           W   
1183  Steamrollers       67    1245.95    1237.63             A           W   
...            ...      ...        ...        ...           ...         ...   
9701         Hawks       93    1564.97    1571.43             H           L   
9702         Hawks      101    1571.43    1566.48             H           W   
9709         Hawks      110    1566.48    1551.37             A           W   
9711         Hawks      117    1551.37    1570.45             H           L   
9715         Hawks       97    1570.45    1574.90             A           L   

      forecast notes  
1136      0.39   NaN  
1152      0.44   NaN  
1159      0.61   NaN  
1170      0.43   NaN  
1183      0.74   NaN  
...        ...   ...  
9701      0.51   NaN  
9702      0.49   NaN  
9709      0.24   NaN  
9711      0.55   NaN  
9715      0.24   NaN  

[946 rows x 23 columns]

In [29]:
df_test=df.loc[df["team_id"] == "MNL"]
df_test["date_game"]

1136     11/4/1948
1152     11/6/1948
1159     11/9/1948
1170    11/11/1948
1183    11/13/1948
           ...    
9701     3/19/1960
9702     3/20/1960
9709     3/22/1960
9711     3/24/1960
9715     3/26/1960
Name: date_game, Length: 946, dtype: object

In [39]:
df.loc[df["team_id"] == "MNL", "date_played"].agg(("min", "max"))

min   1948-11-04
max   1960-03-26
Name: date_played, dtype: datetime64[ns]

Lakers played between the years of 1948 and 1960
<br>
How many points the Boston Celtics have scored during all matches contained in this dataset

In [40]:
df.loc[df["team_id"] == "BOS", "pts"].sum()

626484

### Getting to Know Pandas’ Data Structures
DataFrame provides functions that can feel quite intuitive

### Understanding Series Objects


In [41]:
revenues = pd.Series([5555, 7000, 1980])
revenues

0    5555
1    7000
2    1980
dtype: int64

A Series object wraps two components:

* A sequence of values
* A sequence of identifiers, which is the index
<br>
You can access these components with .values and .index, respectively:



In [43]:
revenues.values

array([5555, 7000, 1980], dtype=int64)

In [44]:
revenues.index

RangeIndex(start=0, stop=3, step=1)

Series can also have an arbitrary type of index

In [45]:
city_revenues = pd.Series(
     [4200, 8000, 6500],
     index=["Amsterdam", "Toronto", "Tokyo"]
 )

In [126]:
type(city_revenues)

pandas.core.series.Series

In [46]:
city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

You can use the code blocks above to distinguish between two types of Series:

* revenues: This Series behaves like a Python list because it only has a positional index.
* city_revenues: This Series acts like a Python dictionary because it features both a positional and a label index.
<br>
construct a Series with a label index from a Python dictionary:

In [47]:
city_employee_count = pd.Series({"Amsterdam": 5, "Tokyo": 8})
city_employee_count

Amsterdam    5
Tokyo        8
dtype: int64

The dictionary keys become the index, and the dictionary values are the Series values.

In [48]:
city_employee_count.keys()

Index(['Amsterdam', 'Tokyo'], dtype='object')

In [49]:
"Tokyo" in city_employee_count

True

In [50]:
"New York" in city_employee_count

False

### Understanding DataFrame Objects
* Series is a pretty powerful data structure, it has its limitations.
* For example, you can only store one attribute per key. As you’ve seen with the nba dataset, which features 23 columns, the Pandas Python library has more to offer with its DataFrame. 
* This data structure is a sequence of Series objects that share the same index

In [130]:
city_employee_count

Amsterdam    5
Tokyo        8
dtype: int64

In [51]:
city_data = pd.DataFrame({
     "revenue": city_revenues,#[4200,8000..]
     "employee_count": city_employee_count
 })

In [52]:
city_data

revenue  employee_count
Amsterdam     4200             5.0
Tokyo         6500             8.0
Toronto       8000             NaN

In [53]:
city_data.duplicates().sum()

Index(['Amsterdam', 'Tokyo', 'Toronto'], dtype='object')

In [54]:
city_data.values

array([[4.2e+03, 5.0e+00],
       [6.5e+03, 8.0e+00],
       [8.0e+03,     nan]])

In [55]:
city_data.axes[0]

Index(['Amsterdam', 'Tokyo', 'Toronto'], dtype='object')

In [56]:
city_data.axes[1]

Index(['revenue', 'employee_count'], dtype='object')

axis marked with 0 is the row index, and the axis marked with 1 is the column index
<br>
DataFrame is also a dictionary-like data structure, so it also supports .keys() 

In [58]:
city_data.keys()

Index(['revenue', 'employee_count'], dtype='object')

In [59]:
"revenue" in city_data

True

### Accessing Series Elements
two Pandas-specific access methods:

.loc
.iloc

### Using the Indexing Operator
Recall that a Series has two indices:

* A positional or implicit index, which is always a RangeIndex
* A label or explicit index, which can contain any hashable objects


In [60]:
city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

In [61]:
city_revenues["Toronto"]

8000

In [62]:
city_revenues[1]

8000

You can also use negative indices and slices, just like you would for a list:



In [63]:
city_revenues[-1]

6500

In [64]:
city_revenues[1:]

Toronto    8000
Tokyo      6500
dtype: int64

In [65]:
city_revenues["Toronto":]

Toronto    8000
Tokyo      6500
dtype: int64

### Using .loc and .iloc


In [66]:
colors = pd.Series(
     ["red", "purple", "blue", "green", "yellow"],
     index=[1, 2, 3, 5, 8]
 )

In [67]:
colors

1       red
2    purple
3      blue
5     green
8    yellow
dtype: object

For a positional index, colors[1] is "purple". However, if you go by the label index, then colors[1] is referring to "red"

Data access methods:

* .loc refers to the label index.
* .iloc refers to the positional index

In [68]:
colors.loc[1]

'red'

In [69]:
colors.iloc[1]

'purple'

colors.loc[1] returned "red", the element with the label 1. colors.iloc[1] returned "purple", the element with the index 1

In [73]:
colors.iloc[1:3]

2    purple
3      blue
dtype: object

In [74]:
colors.iloc[-2]

'green'

In [75]:
colors

1       red
2    purple
3      blue
5     green
8    yellow
dtype: object

* You can use .iloc on a Series similar to using [] on a list.
* You can use .loc on a Series similar to using [] on a dictionary.

### Accessing DataFrame Elements
DataFrame as a dictionary whose values are Series

In [77]:
city_data["revenue"]

Amsterdam    4200
Tokyo        6500
Toronto      8000
Name: revenue, dtype: int64

In [78]:
# Only when col name is string 
city_data.revenue

Amsterdam    4200
Tokyo        6500
Toronto      8000
Name: revenue, dtype: int64

Accessing DataFrame elements with dot notation may not work or may lead to surprises. This is when a column name coincides with a DataFrame attribute or method name

In [81]:
toys = pd.DataFrame([
     {"name": "ball", "shape": "sphere"},
     {"name": "Rubik's cube", "shape": "cube"}
 ])
toys["shape"]

0    sphere
1      cube
Name: shape, dtype: object

In [132]:
city_data

revenue  employee_count
Amsterdam     4200             5.0
Tokyo         6500             8.0
Toronto       8000             NaN

In [82]:
### Using .loc and .iloc
city_data.loc["Amsterdam"]

revenue           4200.0
employee_count       5.0
Name: Amsterdam, dtype: float64

In [83]:
city_data.loc["Tokyo": "Toronto"]

revenue  employee_count
Tokyo       6500             8.0
Toronto     8000             NaN

In [84]:
city_data.iloc[1]

revenue           6500.0
employee_count       8.0
Name: Tokyo, dtype: float64

* city_data.loc["Amsterdam"] selects the row with the label index "Amsterdam".
* city_data.loc["Tokyo": "Toronto"] selects the rows with label indices from "Tokyo" to "Toronto". Remember, .loc is inclusive.
* city_data.iloc[1] selects the row with the positional index 1, which is "Tokyo"

In [85]:
df.iloc[-2]

gameorder                      63157
game_id                 201506170CLE
lg_id                            NBA
_iscopy                            0
year_id                         2015
date_game                  6/16/2015
seasongame                       102
is_playoffs                        1
team_id                          CLE
fran_id                    Cavaliers
pts                               97
elo_i                        1700.74
elo_n                        1692.09
win_equiv                      59.29
opp_id                           GSW
opp_fran                    Warriors
opp_pts                          105
opp_elo_i                    1813.63
opp_elo_n                    1822.29
game_location                      H
game_result                        L
forecast                        0.48
notes                            NaN
date_played      2015-06-16 00:00:00
Name: 126312, dtype: object

You can use these parameters together to select a subset of rows and columns from your DataFrame

In [86]:
city_data.loc["Amsterdam": "Tokyo", "revenue"]

Amsterdam    4200
Tokyo        6500
Name: revenue, dtype: int64

In [87]:
df.loc[5555:5559, ["fran_id", "opp_fran", "pts", "opp_pts"]]

fran_id  opp_fran  pts  opp_pts
5555  Pistons  Warriors   83       56
5556  Celtics    Knicks   95       74
5557   Knicks   Celtics   74       95
5558    Kings    Sixers   81       86
5559   Sixers     Kings   86       81

### Querying Your Dataset
Select rows based on the values in your dataset’s columns to query your data

In [90]:
# only games played after 2010
current_decade = df[df["year_id"] > 2010]
current_decade.shape

(12658, 24)

You now have 24 columns, but your new DataFrame only consists of rows where the value in the "year_id" column is greater than 2010.
<br>
You can also select the rows where a specific field is not null:





In [91]:
games_with_notes = df[df["notes"].notnull()]
games_with_notes.shape
#isna()

(5424, 24)

You can even access values of the object data type as str and perform string methods on them:



In [92]:
ers = df[df["fran_id"].str.endswith("ers")]
ers.shape

(27797, 24)

You use .str.endswith() to filter your dataset and find all games where the home team’s name ends with "ers".



In [94]:
# multiple criteria and query your dataset
df[
     (df["_iscopy"] == 0) & # 
     (df["pts"] > 100) &
     (df["opp_pts"] > 100) &
     (df["team_id"] == "BLB")
 ]

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
1726        864  194902260BLB   NBA        0     1949  2/26/1949          53   
4890       2446  195301100BLB   NBA        0     1953  1/10/1953          32   
4909       2455  195301140BLB   NBA        0     1953  1/14/1953          34   
5208       2605  195303110BLB   NBA        0     1953  3/11/1953          66   
5825       2913  195402220BLB   NBA        0     1954  2/22/1954          60   

      is_playoffs team_id    fran_id  pts    elo_i    elo_n  win_equiv opp_id  \
1726            0     BLB  Baltimore  114  1421.94  1419.43      38.56    MNL   
4890            0     BLB  Baltimore  126  1328.67  1356.65      25.80    BOS   
4909            0     BLB  Baltimore  104  1349.83  1346.36      24.88    MNL   
5208            0     BLB  Baltimore  107  1284.52  1282.24      19.58    NYK   
5825            0     BLB  Baltimore  110  1303.75  1301.97      20.74    BOS   

     opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
1726   Lakers      115    1637.99    1640.49             H           L   
4890  Celtics      105    1591.14    1563.17             H           W   
4909   Lakers      112    1665.44    1668.91             H           L   
5208   Knicks      113    1649.15    1651.44             H           L   
5825  Celtics      111    1591.49    1593.27             H           L   

      forecast            notes date_played  
1726      0.34              NaN  1949-02-26  
4890      0.28              NaN  1953-01-10  
4909      0.22              NaN  1953-01-14  
5208      0.18     at Boston MA  1953-03-11  
5825      0.25  at Worcester MA  1954-02-22

df["_iscopy"] == 0 to include only the entries that aren’t copies

In [95]:
df[
     (df["_iscopy"] == 0) &
     (df["team_id"].str.startswith("LA")) &
     (df["year_id"]==1992) &
     (df["notes"].notnull())
 ]

gameorder       game_id lg_id  _iscopy  year_id date_game  seasongame  \
68901      34451  199205030LAC   NBA        0     1992  5/3/1992          86   
68903      34452  199205030LAL   NBA        0     1992  5/3/1992          86   

       is_playoffs team_id   fran_id  pts    elo_i   elo_n  win_equiv opp_id  \
68901            1     LAC  Clippers  115  1546.90  1556.0      47.89    UTA   
68903            1     LAL    Lakers   76  1489.65  1474.4      39.40    POR   

           opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
68901          Jazz      107    1647.51    1638.41             H           W   
68903  Trailblazers      102    1655.12    1670.37             H           L   

       forecast                                       notes date_played  
68901      0.50  at Anaheim, CA (Anaheim Convention Center)  1992-05-03  
68903      0.41                            at Las Vegas, NV  1992-05-03

In [ ]:
def get_grades(value):
    if value > 70:
        return 'Very good'
    elif 60<value<70 :
        return 'good'
    else: 
        return 'bad'
df['Grades']=df['Grades'].apply(get_grades)
.agg

### Grouping and Aggregating Your Data
grouping and aggregation functions like mean, avg

In [177]:
city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

In [178]:
type(city_revenues)

pandas.core.series.Series

In [96]:
city_revenues.sum()

18700

In [98]:
city_revenues.max()

8000

In [99]:
points = df["pts"]
type(points)


pandas.core.series.Series

In [100]:
df["pts"].sum()

12976235

A DataFrame can have multiple columns, which introduces new possibilities for aggregations, like grouping:



In [101]:
df.groupby("fran_id", sort=False)["pts"].sum()

fran_id
Huskies           3995
Knicks          582497
Stags            20398
Falcons           3797
Capitols         22387
Celtics         626484
Steamrollers     12372
Ironmen           3674
Bombers          17793
Rebels            4474
Warriors        591224
Baltimore        37219
Jets              4482
Pistons         572758
Lakers          637444
Kings           569245
Hawks           567261
Denver            4818
Olympians        22864
Redskins          5372
Waterloo          4921
Packers           6193
Sixers          585891
Wizards         474809
Bulls           437269
Thunder         437735
Squires          91127
Stars            84940
Rockets         432504
Colonels         94435
Pacers          438288
Nuggets         445780
Spurs           453822
Spirits          85874
Sounds           75582
Floridians       49568
Nets            417809
Condors          49642
Bucks           418326
Suns            437486
Clippers        380523
Cavaliers       380416
Trailblazers    402695
Sai

In [102]:
# group by multiple columns:
df[
     (df["fran_id"] == "Spurs") &
     (df["year_id"] > 2010)
 ].groupby(["year_id", "game_result"])["game_id"].count()

year_id  game_result
2011     L              25
         W              63
2012     L              20
         W              60
2013     L              30
         W              73
2014     L              27
         W              78
2015     L              31
         W              58
Name: game_id, dtype: int64

In [103]:
df[
     (df["fran_id"] == "Warriors") &
     (df["year_id"] == 2015)
 ].groupby(["is_playoffs", "game_result"])["game_id"].count()

is_playoffs  game_result
0            L              15
             W              67
1            L               5
             W              16
Name: game_id, dtype: int64

### Manipulating Columns


In [104]:
# Create a copy of your original DataFrame to work with:
df_copy=df.copy()
df_copy.shape

(126314, 24)

In [105]:
df_copy["difference"] = df_copy.pts - df_copy.opp_pts # df["pts"] df["opp  pts"]
df_copy.shape

(126314, 25)

In [106]:
df_copy["difference"].max()

68

In [ ]:
df.rename(
     columns={"game_result": "result", "game_location": "location"},
 )

In [107]:
renamed_df = df.rename(
     columns={"game_result": "result", "game_location": "location"},inplace=True
 )
renamed_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 24 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   gameorder    126314 non-null  int64         
 1   game_id      126314 non-null  object        
 2   lg_id        126314 non-null  object        
 3   _iscopy      126314 non-null  int64         
 4   year_id      126314 non-null  int64         
 5   date_game    126314 non-null  object        
 6   seasongame   126314 non-null  int64         
 7   is_playoffs  126314 non-null  int64         
 8   team_id      126314 non-null  object        
 9   fran_id      126314 non-null  object        
 10  pts          126314 non-null  int64         
 11  elo_i        126314 non-null  float64       
 12  elo_n        126314 non-null  float64       
 13  win_equiv    126314 non-null  float64       
 14  opp_id       126314 non-null  object        
 15  opp_fran     126314 non-null  obje

In [108]:
elo_columns = ["elo_i", "elo_n", "opp_elo_i", "opp_elo_n"]
df_copy.drop(elo_columns, inplace=True, axis=1)
df_copy.shape

(126314, 21)

### Specifying Data Types
Pandas assigns a data type to each column based on its values. While it does a pretty good job, it’s not perfect. If you choose the right data type for your columns

In [109]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 24 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   gameorder      126314 non-null  int64         
 1   game_id        126314 non-null  object        
 2   lg_id          126314 non-null  object        
 3   _iscopy        126314 non-null  int64         
 4   year_id        126314 non-null  int64         
 5   date_game      126314 non-null  object        
 6   seasongame     126314 non-null  int64         
 7   is_playoffs    126314 non-null  int64         
 8   team_id        126314 non-null  object        
 9   fran_id        126314 non-null  object        
 10  pts            126314 non-null  int64         
 11  elo_i          126314 non-null  float64       
 12  elo_n          126314 non-null  float64       
 13  win_equiv      126314 non-null  float64       
 14  opp_id         126314 non-null  object        
 15  

In [110]:
df["date_game"] = pd.to_datetime(df["date_game"])

dtype('<M8[ns]')

In [111]:
df["game_location"] = pd.Categorical(df["game_location"])
df["game_location"].dtype

CategoricalDtype(categories=['A', 'H', 'N'], ordered=False)

* categorical data has a few advantages over unstructured text. 
* When you specify the categorical data type, you make validation easier and save a ton of memory
* Pandas will only use the unique values internally

In [112]:
print(df["game_result"].nunique())
print(df["game_result"].value_counts())

2
L    63157
W    63157
Name: game_result, dtype: int64


In [113]:
df["game_result"] = pd.Categorical(df["game_result"])

In [114]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 24 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   gameorder      126314 non-null  int64         
 1   game_id        126314 non-null  object        
 2   lg_id          126314 non-null  object        
 3   _iscopy        126314 non-null  int64         
 4   year_id        126314 non-null  int64         
 5   date_game      126314 non-null  datetime64[ns]
 6   seasongame     126314 non-null  int64         
 7   is_playoffs    126314 non-null  int64         
 8   team_id        126314 non-null  object        
 9   fran_id        126314 non-null  object        
 10  pts            126314 non-null  int64         
 11  elo_i          126314 non-null  float64       
 12  elo_n          126314 non-null  float64       
 13  win_equiv      126314 non-null  float64       
 14  opp_id         126314 non-null  object        
 15  

As you work with more massive datasets, memory savings becomes especially crucial. Be sure to keep performance in mind as you continue to explore your datasets.



## NumPy
Provides a simple yet powerful data structure: the n-dimensional array
### Benefits
* More speed: NumPy uses algorithms written in C that complete in nanoseconds rather than seconds.
* Fewer loops: NumPy helps you to reduce loops and keep from getting tangled up in iteration indices.
* Clearer code: Without loops, your code will look more like the equations you’re trying to calculate.
* Better quality: There are thousands of contributors working to keep NumPy fast, friendly, and bug free.

In [133]:
import numpy as np

In [134]:
digits = np.array([
        [1, 2, 3],
        [4, 5, 6],
        [6, 7, 9],
    ])
digits

array([[1, 2, 3],
       [4, 5, 6],
       [6, 7, 9]])

In [135]:
# Check type
type(digits)

numpy.ndarray

#### Numpy Example
* Creating arrays using numpy.array()
* Treating complete arrays like individual values to make vectorized calculations more readable
* Using built-in NumPy functions to modify and aggregate the data


In [136]:
# Define an array of grades 
grades = np.array([72, 35, 64, 88, 51, 90, 74, 12])


In [138]:
# Define funciton to do curve for grades
CURVE_CENTER = 80
def curve(grades):
     average = grades.mean()
     change = CURVE_CENTER - average
     new_grades = grades + change
     return np.clip(new_grades, grades, 100)

curve(grades)

array([ 91.25,  54.25,  83.25, 100.  ,  70.25, 100.  ,  93.25,  31.25])

* Creates NumPy array, which is one-dimensional and has a shape of (8,) and a data type of int64
* Takes the average of all the scores using .mean()
<br>
<b>Advantage of two important concepts at once:</b>

* Vectorization
* Broadcasting

<b>Vectorization</b> is the process of performing the same operation in the same way for each element in an array. This removes for loops from your code but achieves the same result.
<br>
<b>Broadcasting</b> is the process of extending two arrays of different shapes and figuring out how to perform a vectorized calculation between them. Remember, grades is an array of numbers of shape (8,) and change is a scalar, or single number, essentially with shape (1,)


### Numpy Shape
* Shape is a key concept when you’re using multidimensional arrays
* Arrays have a property called .shape that returns a tuple of the size in each dimension

In [139]:
temperatures = np.array([
        29.3, 42.1, 18.8, 16.1, 38.0, 12.5,
        12.6, 49.9, 38.6, 31.3, 9.2, 22.2
    ]).reshape(2, 2, 3)
temperatures

array([[[29.3, 42.1, 18.8],
        [16.1, 38. , 12.5]],

       [[12.6, 49.9, 38.6],
        [31.3,  9.2, 22.2]]])

In [140]:
temperatures.shape

(2, 2, 3)

In [142]:
np.swapaxes(temperatures, 1, 2)

array([[[29.3, 16.1],
        [42.1, 38. ],
        [18.8, 12.5]],

       [[12.6, 31.3],
        [49.9,  9.2],
        [38.6, 22.2]]])

* numpy.ndarray method called .reshape() to form a 2 × 2 × 3 block of data
* printed arrays quickly become hard to visualize in three or more dimensions
* swap axes with .swapaxes(), it becomes little clearer which dimension is which

## Axes
* Not only what shape your data is in but also which data is in which axis
* Two-dimensional array has a vertical axis (axis 0) and a horizontal axis (axis 1)
* Lots of functions and commands in NumPy change their behavior based on which axis you tell them to process

In [143]:
table = np.array([
        [5, 3, 7, 1],
        [2, 6, 7 ,9],
        [1, 1, 1, 1],
        [4, 3, 2, 0],
    ])

In [144]:
table.max()

9

In [145]:
table.max(axis=0)

array([5, 6, 7, 9])

In [146]:
table.max(axis=1)

array([7, 9, 1, 4])

* max() returns the largest value in the entire array, no matter how many dimensions there are
* argument of axis=0, .max() selects the maximum value in each of the four vertical sets
<br>
<b>If no axis is specified, then they perform an operation on the entire dataset

### Broadcasting
Arrays can be broadcast against each other if their dimensions match or if one of the arrays has a size of 1

In [147]:
A = np.arange(32).reshape(4, 1, 8)
A

array([[[ 0,  1,  2,  3,  4,  5,  6,  7]],

       [[ 8,  9, 10, 11, 12, 13, 14, 15]],

       [[16, 17, 18, 19, 20, 21, 22, 23]],

       [[24, 25, 26, 27, 28, 29, 30, 31]]])

In [149]:
B = np.arange(48).reshape(1, 6, 8)
B

array([[[ 0,  1,  2,  3,  4,  5,  6,  7],
        [ 8,  9, 10, 11, 12, 13, 14, 15],
        [16, 17, 18, 19, 20, 21, 22, 23],
        [24, 25, 26, 27, 28, 29, 30, 31],
        [32, 33, 34, 35, 36, 37, 38, 39],
        [40, 41, 42, 43, 44, 45, 46, 47]]])

A has 4 planes, each with 1 row and 8 columns. B has only 1 plane with 6 rows and 8 columns

In [150]:
A+B

array([[[ 0,  2,  4,  6,  8, 10, 12, 14],
        [ 8, 10, 12, 14, 16, 18, 20, 22],
        [16, 18, 20, 22, 24, 26, 28, 30],
        [24, 26, 28, 30, 32, 34, 36, 38],
        [32, 34, 36, 38, 40, 42, 44, 46],
        [40, 42, 44, 46, 48, 50, 52, 54]],

       [[ 8, 10, 12, 14, 16, 18, 20, 22],
        [16, 18, 20, 22, 24, 26, 28, 30],
        [24, 26, 28, 30, 32, 34, 36, 38],
        [32, 34, 36, 38, 40, 42, 44, 46],
        [40, 42, 44, 46, 48, 50, 52, 54],
        [48, 50, 52, 54, 56, 58, 60, 62]],

       [[16, 18, 20, 22, 24, 26, 28, 30],
        [24, 26, 28, 30, 32, 34, 36, 38],
        [32, 34, 36, 38, 40, 42, 44, 46],
        [40, 42, 44, 46, 48, 50, 52, 54],
        [48, 50, 52, 54, 56, 58, 60, 62],
        [56, 58, 60, 62, 64, 66, 68, 70]],

       [[24, 26, 28, 30, 32, 34, 36, 38],
        [32, 34, 36, 38, 40, 42, 44, 46],
        [40, 42, 44, 46, 48, 50, 52, 54],
        [48, 50, 52, 54, 56, 58, 60, 62],
        [56, 58, 60, 62, 64, 66, 68, 70],
        [64, 66, 68, 70, 72,

* NumPy duplicates the plane in B three times so that you have a total of four, matching the number of planes in A. 
* It also duplicates the single row in A five times for a total of six, matching the number of rows in B

## Filter, Order, Aggregate
### Indexing
Positive or negative indices to index from the front or back of the array
<br>
Difference with list NumPy arrays use commas between axes, so you can index multiple axes in one set of square brackets

In [160]:
# Basic indexing 
import numpy as np
a = np.arange(20)
print(a)
a[-8:17:1]

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


array([12, 13, 14, 15, 16])

In [161]:
a[10:]

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])

In [163]:
b = np.array([[[1, 2, 3],[4, 5, 6]],
            [[7, 8, 9],[10, 11, 12]]])
b

array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]]])

In [164]:
b.shape

(2, 2, 3)

In [165]:
b[...,1]

array([[ 2,  5],
       [ 8, 11]])

### Masking

In [166]:
numbers = np.linspace(5, 50, 24, dtype=int).reshape(4, -1)
numbers

array([[ 5,  6,  8, 10, 12, 14],
       [16, 18, 20, 22, 24, 26],
       [28, 30, 32, 34, 36, 38],
       [40, 42, 44, 46, 48, 50]])

In [167]:
mask = numbers % 4 == 0

In [168]:
numbers[mask]

array([ 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48])

In [170]:
by_four = numbers[numbers % 4 == 0]
by_four

array([ 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48])

### Transposing

In [171]:
a= np.array([
        [1, 2],
        [3, 4],
        [5, 6],
    ])
a

array([[1, 2],
       [3, 4],
       [5, 6]])

In [172]:
a.T

array([[1, 3, 5],
       [2, 4, 6]])

In [173]:
data = np.array([
        [7, 1, 4],
        [8, 6, 5],
        [1, 2, 3]
    ])
data

array([[7, 1, 4],
       [8, 6, 5],
       [1, 2, 3]])

In [174]:
np.sort(data)

array([[1, 4, 7],
       [5, 6, 8],
       [1, 2, 3]])

In [175]:
np.sort(data, axis=None)

array([1, 1, 2, 3, 4, 5, 6, 7, 8])

In [176]:
np.sort(data, axis=0)

array([[1, 1, 3],
       [7, 2, 4],
       [8, 6, 5]])

In [ ]:

isna()